File: android-native-app/app/src/main/java/pl/edu/pb/androidnativeapp/data/infinitelist/repository/InfiniteListRepositoryImpl.kt
package pl.edu.pb.androidnativeapp.data.infinitelist.repository

import pl.edu.pb.androidnativeapp.data.infinitelist.remote.InfiniteListDataSource
import pl.edu.pb.androidnativeapp.model.Character
import pl.edu.pb.androidnativeapp.model.Characters

class InfiniteListRepositoryImpl(
    private val infiniteListDataSource: InfiniteListDataSource,
) : InfiniteListRepository {
    override suspend fun getCharacters(url: String): Characters {
        val charactersPage = infiniteListDataSource.getCharactersPage(url)
        return Characters(
            characters = charactersPage.results.map {
                Character(
                    id = it.id,
                    name = it.name,
                    status = it.status,
                    species = it.species,
                    gender = it.gender,
                    image = it.image,
                )
            },
            nextPageUrl = charactersPage.info.next,
        )
    }

    override suspend fun getAllCharacters(): List<Character> {
        var url: String? = "https://rickandmortyapi.com/api/character"
        val allCharacters = mutableListOf<Character>()
        do {
            val charactersPage = infiniteListDataSource.getCharactersPage(url!!)
            allCharacters += charactersPage.results.map {
                Character(
                    id = it.id,
                    name = it.name,
                    status = it.status,
                    species = it.species,
                    gender = it.gender,
                    image = it.image,
                )
            }
            url = charactersPage.info.next
        } while (url != null)
        return allCharacters
    }
}

File: flutter-multiplatform-app/flutter_multiplatform_app/lib/data/infinitylist/repository/infinite_list_repository_impl.dart
import 'package:flutter_multiplatform_app/data/infinitylist/remote/infinite_list_data_source.dart';
import 'package:flutter_multiplatform_app/data/infinitylist/repository/infinite_list_repository.dart';
import 'package:flutter_multiplatform_app/model/character.dart';
import 'package:flutter_multiplatform_app/model/characters.dart';

class InfiniteListRepositoryImpl implements InfiniteListRepository {
  InfiniteListRepositoryImpl({required InfiniteListDataSource infiniteListDataSource})
      : _infiniteListDataSource = infiniteListDataSource;

  final InfiniteListDataSource _infiniteListDataSource;

  @override
  Future<Characters> getCharacters(String url) async {
    final charactersPage = await _infiniteListDataSource.getCharactersPage(url);
    return Characters(
        nextPageUrl: charactersPage.info.next,
        characters: charactersPage.results.map((result) {
          return Character(
              id: result.id,
              name: result.name,
              status: result.status,
              species: result.species,
              gender: result.gender,
              image: result.image);
        }).toList());
  }

  @override
  Future<List<Character>> getAllCharacters() async {
    String? url = "https://rickandmortyapi.com/api/character";
    late List<Character> allCharacters = [];
    do {
      final charactersPage = await _infiniteListDataSource.getCharactersPage(url!);
      allCharacters += charactersPage.results.map((result) {
        return Character(
            id: result.id,
            name: result.name,
            status: result.status,
            species: result.species,
            gender: result.gender,
            image: result.image);
      }).toList();
      url = charactersPage.info.next;
    } while (url != null);
    return allCharacters;
  }
}
